File: packages/on_audio_query/on_audio_query_android/android/src/main/kotlin/com/lucasjosino/on_audio_query/methods/queries/AudiosQuery.kt
package com.lucasjosino.on_audio_query.methods.queries

import android.annotation.SuppressLint
import android.content.ContentResolver
import android.content.Context
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.lucasjosino.on_audio_query.controllers.PermissionController
import com.lucasjosino.on_audio_query.methods.helper.QueryHelper
import com.lucasjosino.on_audio_query.types.checkAudioType
import com.lucasjosino.on_audio_query.types.checkAudiosUriType
import com.lucasjosino.on_audio_query.types.sorttypes.checkAudioSortType
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/** AudiosQuery */
class AudiosQuery : ViewModel() {

    // Main parameters
    private val helper = QueryHelper()
    private var selection: String = ""

    // None of this methods can be null.
    private lateinit var uri: Uri
    private lateinit var resolver: ContentResolver
    private lateinit var sortType: String

    // Audios projection
    // Ignore the [Data] deprecation because this plugin support older versions.
    @Suppress("DEPRECATION")
    val audioProjection: Array<String>
        @SuppressLint("InlinedApi")
        get() : Array<String> {
            val tmpProjection = arrayListOf(
                MediaStore.Audio.Media._ID,
                MediaStore.Audio.Media.DATA,
                MediaStore.Audio.Media.DISPLAY_NAME,
                MediaStore.Audio.Media.SIZE,
                MediaStore.Audio.Media.ALBUM,
                MediaStore.Audio.Media.ALBUM_ARTIST,
                MediaStore.Audio.Media.ALBUM_ID,
                MediaStore.Audio.Media.ARTIST,
                MediaStore.Audio.Media.ARTIST_ID,
                MediaStore.Audio.Media.BOOKMARK,
                MediaStore.Audio.Media.COMPOSER,
                MediaStore.Audio.Media.DATE_ADDED,
                MediaStore.Audio.Media.DATE_MODIFIED,
                MediaStore.Audio.Media.DURATION,
                MediaStore.Audio.Media.TITLE,
                MediaStore.Audio.Media.TRACK,
                MediaStore.Audio.Media.YEAR,
                MediaStore.Audio.Media.IS_ALARM,
                MediaStore.Audio.Media.IS_MUSIC,
                MediaStore.Audio.Media.IS_NOTIFICATION,
                MediaStore.Audio.Media.IS_PODCAST,
                MediaStore.Audio.Media.IS_RINGTONE,
            )

            // Only Api >= 29
            if (Build.VERSION.SDK_INT >= 29) {
                tmpProjection.add(MediaStore.Audio.Media.IS_AUDIOBOOK)
            }

            // Only Api >= 30
            if (Build.VERSION.SDK_INT >= 30) {
                tmpProjection.add(MediaStore.Audio.Media.GENRE)
                tmpProjection.add(MediaStore.Audio.Media.GENRE_ID)
            }

            return tmpProjection.toTypedArray()
        }

    @Suppress("UNCHECKED_CAST")
    fun init(
        context: Context,
        // Call from 'MethodChannel' (method).
        result: MethodChannel.Result? = null,
        call: MethodCall? = null,
        // Call from 'EventChannel' (observer).
        sink: EventChannel.EventSink? = null,
        args: Map<*, *>? = null
    ) {
        // Define the [resolver]. This method is used to call the [query].
        resolver = context.contentResolver

        // Define the [args]. Will be delivered from:
        // [result](From MethodChannel) or [sink](From EventChannel)
        val pArgs: Map<String, Any?> = (args ?: call?.arguments) as Map<String, Any?>

        // Define all 'basic' filters.
        val pSortType: Int? = pArgs["sortType"] as Int?
        val pOrderType: Int = pArgs["orderType"] as Int
        val pIgnoreCase: Boolean = pArgs["ignoreCase"] as Boolean
        val pUri: Int = pArgs["uri"] as Int
        val pLimit: Int? = pArgs["limit"] as Int?

        // Define the [toQuery], [toRemove] and [type] filters.
        val toQuery: Map<Int, ArrayList<String>> = pArgs["toQuery"] as Map<Int, ArrayList<String>>
        val toRemove: Map<Int, ArrayList<String>> = pArgs["toRemove"] as Map<Int, ArrayList<String>>
        val type: Map<Int, Int> = pArgs["type"] as Map<Int, Int>

        // Sort: Type and Order.
        sortType = checkAudioSortType(pSortType, pOrderType, pIgnoreCase)

        // Add a 'query' limit(if not null).
        if (pLimit != null) sortType += " LIMIT $pLimit"

        // Check uri:
        //   * [0]: External.
        //   * [1]: Internal.
        uri = checkAudiosUriType(pUri)

        // TODO: Add a generic toQuery and toRemove builder. This will remove a lot of unnecessary code.
        // For every 'row' from 'toQuery', *keep* the media that contains the 'filter'.
        for ((id: Int, values: ArrayList<String>) in toQuery) {
            for (value in values) {
                // The comparison type: contains
                selection += audioProjection[id] + " LIKE '%" + value + "%' " + "AND "
            }
        }

        // For every 'row' from 'toRemove', *remove* the media that contains the 'filter'.
        for ((id: Int, values: ArrayList<String>) in toRemove) {
            for (value in values) {
                // The comparison type: contains
                selection += audioProjection[id] + " NOT LIKE '%" + value + "%' " + "AND "
            }
        }

        // Add/Remove audio type. E.g: is Music, Notification, Alarm, etc..
        for (audioType in type) {
            selection += checkAudioType(audioType.key) + "=" + "${audioType.value} " + "AND "
        }

        // Remove the 'AND ' keyword from selection.
        selection = selection.removeSuffix("AND ")

        // Request permission status from the 'main' method.
        val hasPermission: Boolean = PermissionController().permissionStatus(context)

        // We cannot 'query' without permission so, throw a PlatformException.
        // Only one 'channel' will be 'functional'. If is null, ignore, if not, send the error.
        if (!hasPermission) {
            // Method from 'EventChannel' (observer)
            sink?.error(
                "403",
                "The app doesn't have permission to read files.",
                "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
            )
            // Method from 'MethodChannel' (method)
            result?.error(
                "403",
                "The app doesn't have permission to read files.",
                "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
            )

            // 'Exit' the function
            return
        }

        // Query everything in background for a better performance.
        viewModelScope.launch {
            // Start 'querying'.
            val resultAudioList: ArrayList<MutableMap<String, Any?>> = loadAudios()

            // After loading the information, send the 'result'.
            sink?.success(resultAudioList)
            result?.success(resultAudioList)
        }
    }

    //Loading in Background
    private suspend fun loadAudios(): ArrayList<MutableMap<String, Any?>> =
        withContext(Dispatchers.IO) {
            // Setup the cursor with [uri], [projection] and [sortType].
            val cursor = resolver.query(uri, audioProjection, selection, null, sortType)

            // Empty list.
            val audioList: ArrayList<MutableMap<String, Any?>> = ArrayList()

            // For each item(audio) inside this "cursor", take one and "format"
            // into a [Map<String, dynamic>].
            while (cursor != null && cursor.moveToNext()) {
                val tempData: MutableMap<String, Any?> = HashMap()
                for (audioMedia in cursor.columnNames) {
                    tempData[audioMedia] = helper.loadAudioItem(audioMedia, cursor)
                }

                //Get a extra information from audio, e.g: extension, uri, etc..
                val tempExtraData = helper.loadAudioExtraInfo(uri, tempData)
                tempData.putAll(tempExtraData)

                audioList.add(tempData)
            }

            // Close cursor to avoid memory leaks.
            cursor?.close()

            // After finish the "query", go back to the "main" thread(You can only call flutter
            // inside the main thread).
            return@withContext audioList
        }
}

//Extras:

// * Query only audio > 60000 ms [1 minute]
// Obs: I don't think is a good idea, some audio "Non music" have more than 1 minute
//query(MediaStore.Audio.Media.EXTERNAL_CONTENT_URI, projection, MediaStore.Audio.Media.DURATION +
// ">= 60000", null, checkAudioSortType(sortType!!))

// * Query audio with limit, used for better performance in tests
//MediaStore.Audio.Media.TITLE + " LIMIT 4"

File: packages/on_audio_query/on_audio_query_ios/ios/Classes/methods/queries/AudiosQuery.swift
import Flutter
import MediaPlayer

class AudiosQuery {
    // Main parameters
    private var args: [String: Any]
    private var result: FlutterResult?
    private var sink: FlutterEventSink?
    
    // Audio projection (to filter).
    private let audioProjection: [String?] = [
        "_id",
        "_data",
        "_display_name",
        nil,
        "album",
        nil,
        "album_id",
        "artist",
        "artist_id",
        nil,
        "composer",
        nil,
        nil,
        nil,
        "title",
        nil,
        nil,
        nil,
        nil,
        nil,
        nil,
        nil,
    ]
    
    init(
        // Call from 'MethodChannel' (method).
        call: FlutterMethodCall? = nil,
        result: FlutterResult? = nil,
        // Call from 'EventChannel' (observer).
        sink: FlutterEventSink? = nil,
        args: [String: Any]? = nil
    ) {
        // Get all arguments inside the map.
        self.args = sink != nil ? args! : (call!.arguments as! [String: Any])
        self.sink = sink
        self.result = result
    }
    
    func queryAudios() {
        // The sortType. If 'nil', will be set as [Title].
        let sortType = args["sortType"] as? Int ?? 0
        
        // Define the query (To match android side, let's call "cursor").
        let cursor: MPMediaQuery = MPMediaQuery.init()
    
        // This filter will avoid audios/songs outside phone library(cloud).
        //
        // Sometimes this filter won't work.
        cursor.addFilterPredicate(MPMediaPropertyPredicate.init(
            value: false,
            forProperty: MPMediaItemPropertyIsCloudItem,
            comparisonType: .equalTo
        ))
        
        // Using native sort from [IOS] you can only use the [Title], [Album] and
        // [Artist]. The others will be sorted 'manually' using [formatAudioList] before
        // sending to Dart.
        cursor.groupingType = checkAudioSortType(sortType: sortType)
        
        // Request permission status from the 'main' method.
        let hasPermission = SwiftOnAudioQueryPlugin().checkPermission()
        
        // We cannot 'query' without permission so, throw a PlatformException.
        // Only one 'channel' will be 'functional'. If is null, ignore, if not, send the error.
        if !hasPermission {
            // Call from 'EventChannel' (observer)
            self.sink?(
                FlutterError.init(
                    code: "403",
                    message: "The app doesn't have permission to read files.",
                    details: "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
                )
            )
            
            // Call from 'MethodChannel' (method)
            self.result?(
                FlutterError.init(
                    code: "403",
                    message: "The app doesn't have permission to read files.",
                    details: "Call the [permissionsRequest] method or install a external plugin to handle the app permission."
                )
            )
            
            // 'Exit' the function
            return
        }
        
        // Define the 'new' items.
        var cursorItems: [MPMediaItem] = []
        
        // Get all defined types.
        let types = args["type"] as! [Int: Int]
        
        // TODO: Can 'MPMediaQuery' support differents [values] to the same [property]?
        // TODO: Add ringtones
        // https://github.com/CQH/iOS-Sounds-and-Ringtones/blob/master/iOS%20Sounds%20and%20Ringtones/DirectoriesTableViewController.swift#L85
        
        // To support others audios type we need a 'workaround'.
        //
        // Define the filter [type], get the audio list, add to the [cursorItems]
        // and remove the filter.
        for (type, _) in types {
            // Define the current filter.
            let currentFilter = MPMediaPropertyPredicate.init(
                value: checkAudioType(sortType: type),
                forProperty: MPMediaItemPropertyMediaType,
                comparisonType: .equalTo
            )
            
            // Add the filter to the 'cursor'.
            cursor.addFilterPredicate(currentFilter)
            
            // Get this list and 'append' to another list.
            cursorItems.append(contentsOf: cursor.items ?? [])
            
            // Remove the filter.
            cursor.removeFilterPredicate(currentFilter)
        }
        
        // Query everything in background for a better performance.
        DispatchQueue.global(qos: .userInitiated).async {
            var listOfAudios: [[String: Any?]] = Array()
            
            // Define the 'query' limit.
            let limit: Int? = self.args["limit"] as? Int
            
            // For each item(audio) inside this "cursor", take one and "format"
            // into a [Map<String, dynamic>], all keys are based on [Android]
            // platforms so, if you change some key, will have to change the [Android] too.
            for audio in cursorItems {
                // When list count reach the [limit]. Break the loop.
                //
                // If [limit] value is 'nil', continue.
                if listOfAudios.count == limit {
                    break
                }
                
                // If the audio file don't has a assetURL, is a Cloud item.
                if !audio.isCloudItem && audio.assetURL != nil {
                    let audioData = self.loadAudioItem(audio: audio)
                    listOfAudios.append(audioData)
                }
            }
            
            // Define the [toQuery] and [toRemove] filter.
            let toQuery = self.args["toQuery"] as! [Int: [String]]
            let toRemove = self.args["toRemove"] as! [Int: [String]]
            
            // 'Build' the filter.
            listOfAudios = listOfAudios.mediaFilter(
                mediaProjection: self.audioProjection,
                toQuery: toQuery,
                toRemove: toRemove
            )
            
            // After finish the "query", go back to the "main" thread(You can only call flutter
            // inside the main thread).
            DispatchQueue.main.async {
                // Here we'll check the "custom" sort and define a order to the list.
                let finalList = self.formatAudioList(allAudios: listOfAudios)
                
                // After loading the information, send the 'result'.
                self.sink?(finalList)
                self.result?(finalList)
            }
        }
    }
    
    private func loadAudioItem(audio: MPMediaItem) -> [String: Any?] {
        let fileExt = audio.assetURL?.pathExtension ?? ""
        let sizeInBytes = audio.value(forProperty: "fileSize") as? Int
        return [
            "_id": audio.persistentID,
            "_data": audio.assetURL?.absoluteString,
            "_uri": audio.assetURL?.absoluteString,
            "_display_name": "\(audio.artist ?? "") - \(audio.title ?? "").\(fileExt)",
            "_display_name_wo_ext": "\(audio.artist ?? "") - \(audio.title ?? "")",
            "_size": sizeInBytes,
            "audio_id": nil,
            "album": audio.albumTitle,
            "album_id": audio.albumPersistentID,
            "artist": audio.artist,
            "artist_id": audio.artistPersistentID,
            "genre": audio.genre,
            "genre_id": audio.genrePersistentID,
            "bookmark": Int(audio.bookmarkTime),
            "composer": audio.composer,
            "date_added": Int(audio.dateAdded.timeIntervalSince1970),
            "date_modified": 0,
            "duration": Int(audio.playbackDuration * 1000),
            "title": audio.title,
            "track": audio.albumTrackNumber,
            "file_extension": fileExt,
        ]
    }
    
    private func formatAudioList(allAudios: [[String: Any?]]) -> [[String: Any?]] {
        // Define a copy of all audios.
        var allAudiosCopy = allAudios
        
        // Define all 'basic' filters.
        let order = args["orderType"] as? Int
        let sortType = args["sortType"] as? Int
        let ignoreCase = args["ignoreCase"] as! Bool
        
        // Sort the list 'manually'.
        switch sortType {
        case 3:
            allAudiosCopy.sort { (val1, val2) in
                (val1["duration"] as! Double) > (val2["duration"] as! Double)
            }
        case 4:
            allAudiosCopy.sort { (val1, val2) in
                (val1["date_added"] as! Int) > (val2["date_added"] as! Int)
            }
        case 5:
            allAudiosCopy.sort { (val1, val2) in
                (val1["_size"] as! Int) > (val2["_size"] as! Int)
            }
        case 6:
            allAudiosCopy.sort { (val1, val2) in
                ((val1["_display_name"] as! String).isCase(ignoreCase: ignoreCase)) > ((val2["_display_name"] as! String).isCase(ignoreCase: ignoreCase))
            }
        case 7:
            allAudiosCopy.sort { (val1, val2) in
                (val1["track"] as! Int) > (val2["track"] as! Int)
            }
        default:
            break
        }
        
        // The order value is [1], reverse the list.
        return order == 1 ? allAudiosCopy.reversed() : allAudiosCopy
    }
}
